### PR TITLE
Fixes incorrect parameters in stored templates

### LINKED ISSUE #1713: Nested steps run with incorrect parameters
**Is this a BUG REPORT or FEATURE REQUEST?**:
Bug report

**What happened**:
If you execute the parallelism-nested.yaml example then the pods with a seq-id of b, c or d all have a parallel-id of 1

**What you expected to happen**:
pods with a seq-id of b, c or d all to have all parallel ids (1, 2, 3 and 4).

**How to reproduce it (as minimally and precisely as possible)**:
```argo submit parallelism-nested.yaml```
then:
```argo get <workflow-id> -o yaml```

There will be multiple pods in the Status section with seq-id of b and parallel id of 1, there should only be one.

**Anything else we need to know?**:
This looks to be related to https://github.com/argoproj/argo/pull/1552

storedTemplates contains the following, with seq-id hardcoded:
```
    /seq-worker:
      arguments: {}
      inputs: {}
      metadata: {}
      name: seq-worker
      outputs: {}
      parallelism: 1
      steps:
      - - arguments:
            parameters:
            - name: parallel-id
              value: "1"
            - name: seq-id
              value: '{{item}}'
          name: seq-step
          template: one-job
          withParam: |
            ["a","b","c","d"]
```

**Environment**:
- Argo version: master
```
$ argo version
```
- Kubernetes version : 1.12

### PR DESCRIPTION
**Pull request by simster7**
Fixes #1713. Approach was to store templates _before_ resolution, and to do so all at once during the one-time validation sequence.

Currently not fully tested (hence why this is a Draft PR), just looking for initial feedback on the implementation.

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -161,6 +161,11 @@ func (woc *wfOperationCtx) operate() {
 			woc.markWorkflowFailed(fmt.Sprintf("invalid spec: %s", err.Error()))
 			return
 		}
+		// Store templates
+		for _, template := range woc.wf.Spec.Templates {
+			storedTemplateName := woc.getStoredTemplateName(template.TemplateRef, template.Name, woc.tmplCtx, &wfv1.Template{Template: template.Name})
+			woc.storeTemplate(storedTemplateName, &template)
+		}
```
Comment (simster7) [inline workflow/controller/operator.go]: Store unresolved templates during one-time validation

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -341,22 +346,9 @@ func (woc *wfOperationCtx) getNodeByName(nodeName string) *wfv1.NodeStatus {
...
-				woc.log.Debugf("Switch the template context to %s", node.WorkflowTemplateName)
-				newTmplCtx, err := tmplCtx.WithLazyWorkflowTemplate(woc.wf.Namespace, node.WorkflowTemplateName)
-				if err != nil {
-					return nil, nil, err
-				}
-				return newTmplCtx, resolvedTemplate, nil
-			} else {
-				return tmplCtx.WithTemplateBase(woc.wf), resolvedTemplate, nil
-			}
+		unresolvedTemplate := woc.wf.GetStoredOrLocalTemplate(node)
```
Comment (simster7) [inline workflow/controller/operator.go]: Template will always need to be resolved when querying it.

**Comment by dtaniwaki** [comment]
Thank you for trying to fix the issue. I talked with @jessesuen and @sarabala1979 and concluded that we should remove input/arguments resolving of nested templates and store all the intermediate templates instead because this feature is not expected, then we can simplify the logic and it will hopefully fix all the issues happening right now.

**Comment by simster7** [comment]
@dtaniwaki Are you currently working on the fix?

**Comment by dtaniwaki** [comment]
I’m working on fixing the merging strategy of nested templates.

**Comment by simster7** [comment]
Closed in favor of https://github.com/argoproj/argo/pull/1744

**Pull request closed without merging**